File: amic-platform/src/modules/fdd/types/qoe.ts
// ── Enums ────────────────────────────────────────────────

export type QoEStatus = "DRAFT" | "REVIEW" | "APPROVED";

export type AdjustmentCategory =
  | "NON_RECURRING"
  | "NON_OPERATING"
  | "NORMALIZATION"
  | "OWNER_RELATED"
  | "PRO_FORMA";

export type AdjustmentStatus =
  | "CANDIDATE"
  | "PROPOSED"
  | "APPROVED"
  | "REJECTED";

// ── Adjustment Item ─────────────────────────────────────

export interface AdjustmentItemRead {
  id: string;
  qoe_calculation_id: string;
  deal_id: string;
  category: AdjustmentCategory;
  description: string;
  amount: string; // Decimal as string — NEVER number
  detection_method: string;
  confidence_score: string | null;
  source_account_code: string | null;
  source_account_name: string | null;
  source_entry_ids: string[] | null;
  status: AdjustmentStatus;
  approved_by: string | null;
  approved_at: string | null;
  rejection_reason: string | null;
  display_order: number;
  created_at: string;
  updated_at: string;
}

export interface AdjustmentItemCreate {
  category: AdjustmentCategory;
  description: string;
  amount: string;
  source_account_code?: string | null;
  source_account_name?: string | null;
}

export interface AdjustmentItemUpdate {
  category?: AdjustmentCategory;
  description?: string;
  amount?: string;
  status?: AdjustmentStatus;
  rejection_reason?: string;
}

export interface AdjustmentItemApprove {
  approved_by: string;
}

// ── QoE Calculation ─────────────────────────────────────

export interface QoECalculationRead {
  id: string;
  deal_id: string;
  snapshot_id: string;
  revenue: string;
  cogs: string;
  gross_profit: string;
  sga: string;
  depreciation_amortization: string;
  other_operating: string;
  operating_income: string;
  reported_ebitda: string;
  total_adjustments: string;
  adjusted_ebitda: string;
  balance_check_error: string;
  category_breakdown: Record<string, unknown>;
  engine_version: string;
  status: QoEStatus;
  adjustments: AdjustmentItemRead[];
  created_at: string;
  updated_at: string;
}

export interface QoERunRequest {
  snapshot_id: string;
}

// ── Bridge Summary ──────────────────────────────────────

export interface QoEBridgeSummary {
  reported_ebitda: string;
  adjustments: AdjustmentItemRead[];
  total_adjustments: string;
  adjusted_ebitda: string;
  balance_check_error: string;
  is_balanced: boolean;
}

File: fdd/backend/app/schemas/qoe.py
"""QoE (Quality of Earnings) 관련 Pydantic 스키마 — EPIC-05."""

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.qoe import AdjustmentCategory, AdjustmentStatus, QoEStatus

# ── Adjustment Item ──────────────────────────────────────


class AdjustmentItemCreate(BaseModel):
    """수동 조정항목 추가 요청."""

    category: AdjustmentCategory
    description: str = Field(..., min_length=1, max_length=500)
    amount: Decimal
    source_account_code: str | None = None
    source_account_name: str | None = None


class AdjustmentItemUpdate(BaseModel):
    """조정항목 수정 요청."""

    category: AdjustmentCategory | None = None
    description: str | None = Field(default=None, max_length=500)
    amount: Decimal | None = None
    status: AdjustmentStatus | None = None
    rejection_reason: str | None = None


class AdjustmentItemApprove(BaseModel):
    """조정항목 승인 요청."""

    approved_by: str = Field(..., min_length=1, max_length=100)


class AdjustmentItemRead(BaseModel):
    id: uuid.UUID
    qoe_calculation_id: uuid.UUID
    deal_id: uuid.UUID
    category: AdjustmentCategory
    description: str
    amount: Decimal
    detection_method: str
    confidence_score: Decimal | None
    source_account_code: str | None
    source_account_name: str | None
    source_entry_ids: list | None
    status: AdjustmentStatus
    approved_by: str | None
    approved_at: datetime | None
    rejection_reason: str | None
    display_order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── QoE Calculation ──────────────────────────────────────


class QoERunRequest(BaseModel):
    """QoE 계산 실행 요청."""

    snapshot_id: uuid.UUID


class QoECalculationRead(BaseModel):
    id: uuid.UUID
    deal_id: uuid.UUID
    snapshot_id: uuid.UUID
    revenue: Decimal
    cogs: Decimal
    gross_profit: Decimal
    sga: Decimal
    depreciation_amortization: Decimal
    other_operating: Decimal
    operating_income: Decimal
    reported_ebitda: Decimal
    total_adjustments: Decimal
    adjusted_ebitda: Decimal
    balance_check_error: Decimal
    category_breakdown: dict
    engine_version: str
    status: QoEStatus
    adjustments: list[AdjustmentItemRead]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── QoE Bridge Summary ──────────────────────────────────


class QoEBridgeSummary(BaseModel):
    """QoE Bridge 요약 (프론트엔드 표시용)."""

    reported_ebitda: Decimal
    adjustments: list[AdjustmentItemRead]
    total_adjustments: Decimal
    adjusted_ebitda: Decimal
    balance_check_error: Decimal
    is_balanced: bool
